saleor/graphql/discount/mutations/voucher/voucher_create.py

```
@@ -98,39 +116,151 @@ class Meta:
+                for code_data in codes_data
+            ]
+
+    @classmethod
+    def clean_voucher_instance(cls, info: ResolveInfo, voucher_instance):
+        cls.clean_instance(info, voucher_instance)
+
+        start_date = voucher_instance.start_date
+        end_date = voucher_instance.end_date
+
         try:
             validate_end_is_after_start(start_date, end_date)
         except ValidationError as error:
             error.code = DiscountErrorCode.INVALID.value
             raise ValidationError({"end_date": error})
 
     @classmethod
-    def post_save_action(cls, info: ResolveInfo, instance, cleaned_input):
+    def clean_codes_instance(cls, code_instances):
+        for code_instance in code_instances:
+            code_instance.full_clean(exclude=["voucher"])
+
+    @classmethod
+    def save(cls, _info: ResolveInfo, voucher_instance, code_instances):
+        with transaction.atomic():
+            voucher_instance.save()
+            models.VoucherCode.objects.bulk_create(code_instances)
+
+    @classmethod
+    def post_save_action(cls, info: ResolveInfo, instance, code):
         manager = get_plugin_manager_promise(info.context).get()
-        cls.call_event(manager.voucher_created, instance)
+        cls.call_event(manager.voucher_created, instance, code)
+
+    @classmethod
+    def success_response(cls, instance):
+        instance = ChannelContext(node=instance, channel_slug=None)
+        return super().success_response(instance)
+
+    @classmethod
+    def perform_mutation(cls, _root, info: ResolveInfo, /, **data):
+        voucher_instance = cls.get_instance(info, **data)
+        data = data.get("input")
+        cleaned_input = cls.clean_input(info, voucher_instance, data)
+
+        metadata_list = cleaned_input.pop("metadata", None)
+        private_metadata_list = cleaned_input.pop("private_metadata", None)
+        codes_data = cleaned_input.pop("codes", None)
+        code = cleaned_input.pop("code", None)
```

IMO we should raise an error when both `code` and `codes` fields are provided.